### PR TITLE
suppressor: simplify and reduce locking

### PR DESCRIPTION
**Pull request by jeddenlea**

### CODE REVIEW

File: suppress/suppress.go [hunk 1]
```diff
@@ -43,45 +39,47 @@ func For(duration time.Duration, id string, f func(int, string)) {
...
 
-				_suppressor.Lock()
-				state.lastFunc(state.count, tag)
-				state.next = time.Now().Add(duration)
-				state.count = 0
-				_suppressor.Unlock()
-			}()
+		for range ticker.C {
+			count := atomic.LoadInt64(&state.count)
+			if count == 0 {
```
Comment (twmb) [inline suppress/suppress.go]: This early return doesn't delete the key from the map, meaning nothing will ever be logged for this key again.

Comment (jeddenlea) [inline suppress/suppress.go]: it doesn't need to remove the key from the map, we set count to -1.   That means the next call will trigger it again here: https://github.com/fastly/go-utils/pull/29/files#diff-655af7c92fb732b43708efa830bcda14R57

Comment (twmb) [inline suppress/suppress.go]: ah right

**Comment by twmb** [comment]
lgtm.

**Comment by athomason** [comment]
This breaks all the flaky (but working for me locally on master) tests that are commented out in suppress_test.go.

Old:
```
=== RUN TestSilencer1
--- PASS: TestSilencer1 (1.24s)
	suppress_test.go:81: Ran 275854 iterations in 1.037855964s (1.238174266s with wait), fired correctly 11 times (wanted 11) and 0 incorrectly
=== RUN TestSilencer2
--- PASS: TestSilencer2 (1.23s)
	suppress_test.go:81: Ran 275410 iterations in 1.02647574s (1.227424131s with wait), fired correctly 22 times (wanted 22) and 0 incorrectly
=== RUN TestSilencer3
--- PASS: TestSilencer3 (1.24s)
	suppress_test.go:81: Ran 107475 iterations in 1.041484394s (1.242028826s with wait), fired correctly 33 times (wanted 33) and 0 incorrectly
=== RUN TestSilencer4
--- PASS: TestSilencer4 (1.22s)
	suppress_test.go:81: Ran 96516 iterations in 1.018513677s (1.219018309s with wait), fired correctly 66 times (wanted 66) and 0 incorrectly
=== RUN TestSilencer5
--- PASS: TestSilencer5 (0.21s)
	suppress_test.go:81: Ran 1170 iterations in 11.007245ms (211.698311ms with wait), fired correctly 12 times (wanted 12) and 0 incorrectly
=== RUN TestSilencerStalled
--- PASS: TestSilencerStalled (0.11s)
	suppress_test.go:104: suppress_test.go:105 anon
	suppress_test.go:104: suppress_test.go:105 anon
PASS
ok  	github.com/fastly/go-utils/suppress	5.259s
```

New:
```
=== RUN TestSilencer1
--- FAIL: TestSilencer1 (1.30s)
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:21.718165898 -0700 PDT; delta=99.306287ms attempts=257605 last=2015-03-31 17:37:21.618859242 -0700 PDT count=65301
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:21.918163425 -0700 PDT; delta=99.31468ms attempts=387602 last=2015-03-31 17:37:21.818848297 -0700 PDT count=65475
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:22.118137414 -0700 PDT; delta=99.262938ms attempts=521007 last=2015-03-31 17:37:22.018874197 -0700 PDT count=66515
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:22.318085914 -0700 PDT; delta=99.213168ms attempts=647640 last=2015-03-31 17:37:22.218872395 -0700 PDT count=60059
	suppress_test.go:81: Ran 712935 iterations in 1.100001122s (1.301156699s with wait), fired correctly 8 times (wanted 11) and 4 incorrectly
	suppress_test.go:83: Expected 11 firings, got 8
	suppress_test.go:86: Silencer failed to suppress 4 times
=== RUN TestSilencer2
--- FAIL: TestSilencer2 (1.30s)
	suppress_test.go:43: Error "suppress_test.go:61 spot2" at 2015-03-31 17:37:23.324128883 -0700 PDT; delta=99.994014ms attempts=456355 last=2015-03-31 17:37:23.224134621 -0700 PDT count=33800
	suppress_test.go:43: Error "suppress_test.go:61 spot2" at 2015-03-31 17:37:23.524263215 -0700 PDT; delta=99.988602ms attempts=589008 last=2015-03-31 17:37:23.424274338 -0700 PDT count=32881
	suppress_test.go:81: Ran 704777 iterations in 1.100002187s (1.300302642s with wait), fired correctly 22 times (wanted 22) and 2 incorrectly
	suppress_test.go:86: Silencer failed to suppress 2 times
=== RUN TestSilencer3
--- FAIL: TestSilencer3 (1.30s)
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:24.226099207 -0700 PDT; delta=99.68532ms attempts=65292 last=2015-03-31 17:37:24.126413587 -0700 PDT count=22496
	suppress_test.go:43: Error "suppress_test.go:64 " at 2015-03-31 17:37:24.226146489 -0700 PDT; delta=99.730988ms attempts=65292 last=2015-03-31 17:37:24.1264154 -0700 PDT count=22496
	suppress_test.go:43: Error "suppress_test.go:67 " at 2015-03-31 17:37:24.226163016 -0700 PDT; delta=99.747063ms attempts=65292 last=2015-03-31 17:37:24.126415897 -0700 PDT count=22496
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:24.426103333 -0700 PDT; delta=99.690031ms attempts=111611 last=2015-03-31 17:37:24.326413007 -0700 PDT count=22998
	suppress_test.go:43: Error "suppress_test.go:64 " at 2015-03-31 17:37:24.426152238 -0700 PDT; delta=99.737223ms attempts=111611 last=2015-03-31 17:37:24.326414912 -0700 PDT count=22998
	suppress_test.go:43: Error "suppress_test.go:67 " at 2015-03-31 17:37:24.42616882 -0700 PDT; delta=99.753398ms attempts=111611 last=2015-03-31 17:37:24.326415362 -0700 PDT count=22998
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:24.626107105 -0700 PDT; delta=99.707718ms attempts=155556 last=2015-03-31 17:37:24.52639901 -0700 PDT count=21167
	suppress_test.go:43: Error "suppress_test.go:64 " at 2015-03-31 17:37:24.626359498 -0700 PDT; delta=99.958587ms attempts=155556 last=2015-03-31 17:37:24.526400786 -0700 PDT count=21167
	suppress_test.go:43: Error "suppress_test.go:67 " at 2015-03-31 17:37:24.626377925 -0700 PDT; delta=99.976665ms attempts=155556 last=2015-03-31 17:37:24.526401189 -0700 PDT count=21167
	suppress_test.go:43: Error "suppress_test.go:61 " at 2015-03-31 17:37:24.826112141 -0700 PDT; delta=99.708316ms attempts=200136 last=2015-03-31 17:37:24.72640346 -0700 PDT count=22064
	suppress_test.go:43: Error "suppress_test.go:64 " at 2015-03-31 17:37:24.826185825 -0700 PDT; delta=99.780517ms attempts=200136 last=2015-03-31 17:37:24.726405193 -0700 PDT count=22064
	suppress_test.go:43: Error "suppress_test.go:67 " at 2015-03-31 17:37:24.826201202 -0700 PDT; delta=99.795508ms attempts=200136 last=2015-03-31 17:37:24.726405638 -0700 PDT count=22064
	suppress_test.go:81: Ran 242540 iterations in 1.100002098s (1.300155648s with wait), fired correctly 24 times (wanted 33) and 12 incorrectly
	suppress_test.go:83: Expected 33 firings, got 24
	suppress_test.go:86: Silencer failed to suppress 12 times
=== RUN TestSilencer4
--- FAIL: TestSilencer4 (1.30s)
	suppress_test.go:43: Error "suppress_test.go:67 #2" at 2015-03-31 17:37:25.530479763 -0700 PDT; delta=99.999052ms attempts=55716 last=2015-03-31 17:37:25.430480636 -0700 PDT count=10256
	suppress_test.go:43: Error "suppress_test.go:61 #1" at 2015-03-31 17:37:25.630525194 -0700 PDT; delta=99.999045ms attempts=77467 last=2015-03-31 17:37:25.530526081 -0700 PDT count=10875
	suppress_test.go:43: Error "suppress_test.go:61 #2" at 2015-03-31 17:37:25.830662888 -0700 PDT; delta=97.201092ms attempts=123409 last=2015-03-31 17:37:25.733461441 -0700 PDT count=11192
	suppress_test.go:43: Error "suppress_test.go:64 #2" at 2015-03-31 17:37:25.830870553 -0700 PDT; delta=97.407399ms attempts=123409 last=2015-03-31 17:37:25.733463042 -0700 PDT count=11192
	suppress_test.go:43: Error "suppress_test.go:67 #2" at 2015-03-31 17:37:25.830885858 -0700 PDT; delta=97.42228ms attempts=123409 last=2015-03-31 17:37:25.733463504 -0700 PDT count=11192
	suppress_test.go:43: Error "suppress_test.go:61 #1" at 2015-03-31 17:37:25.830900353 -0700 PDT; delta=97.436247ms attempts=123409 last=2015-03-31 17:37:25.733464038 -0700 PDT count=11191
	suppress_test.go:43: Error "suppress_test.go:64 #1" at 2015-03-31 17:37:25.830911983 -0700 PDT; delta=97.447442ms attempts=123409 last=2015-03-31 17:37:25.733464488 -0700 PDT count=11191
	suppress_test.go:43: Error "suppress_test.go:67 #1" at 2015-03-31 17:37:25.830937527 -0700 PDT; delta=97.472527ms attempts=123409 last=2015-03-31 17:37:25.733464922 -0700 PDT count=11191
	suppress_test.go:43: Error "suppress_test.go:61 #2" at 2015-03-31 17:37:25.930330892 -0700 PDT; delta=99.668139ms attempts=145544 last=2015-03-31 17:37:25.830662533 -0700 PDT count=11067
	suppress_test.go:43: Error "suppress_test.go:64 #2" at 2015-03-31 17:37:25.930376399 -0700 PDT; delta=99.505846ms attempts=145544 last=2015-03-31 17:37:25.830870441 -0700 PDT count=11067
	suppress_test.go:43: Error "suppress_test.go:67 #2" at 2015-03-31 17:37:25.9303908 -0700 PDT; delta=99.504948ms attempts=145544 last=2015-03-31 17:37:25.830885784 -0700 PDT count=11067
	suppress_test.go:43: Error "suppress_test.go:61 #1" at 2015-03-31 17:37:25.930404798 -0700 PDT; delta=99.504455ms attempts=145544 last=2015-03-31 17:37:25.830900285 -0700 PDT count=11068
	suppress_test.go:43: Error "suppress_test.go:64 #1" at 2015-03-31 17:37:25.930416543 -0700 PDT; delta=99.504552ms attempts=145544 last=2015-03-31 17:37:25.83091193 -0700 PDT count=11068
	suppress_test.go:43: Error "suppress_test.go:67 #1" at 2015-03-31 17:37:25.930429234 -0700 PDT; delta=99.491725ms attempts=145544 last=2015-03-31 17:37:25.830937449 -0700 PDT count=11068
	suppress_test.go:43: Error "suppress_test.go:61 #2" at 2015-03-31 17:37:26.130332804 -0700 PDT; delta=99.662969ms attempts=191633 last=2015-03-31 17:37:26.030669487 -0700 PDT count=11599
	suppress_test.go:43: Error "suppress_test.go:64 #2" at 2015-03-31 17:37:26.130382944 -0700 PDT; delta=99.711574ms attempts=191633 last=2015-03-31 17:37:26.030671287 -0700 PDT count=11599
	suppress_test.go:43: Error "suppress_test.go:67 #2" at 2015-03-31 17:37:26.13039745 -0700 PDT; delta=99.725652ms attempts=191633 last=2015-03-31 17:37:26.030671739 -0700 PDT count=11599
	suppress_test.go:43: Error "suppress_test.go:61 #1" at 2015-03-31 17:37:26.130409687 -0700 PDT; delta=99.737428ms attempts=191633 last=2015-03-31 17:37:26.030672202 -0700 PDT count=11598
	suppress_test.go:43: Error "suppress_test.go:64 #1" at 2015-03-31 17:37:26.130422505 -0700 PDT; delta=99.749823ms attempts=191633 last=2015-03-31 17:37:26.030672616 -0700 PDT count=11598
	suppress_test.go:43: Error "suppress_test.go:67 #1" at 2015-03-31 17:37:26.130435415 -0700 PDT; delta=99.762224ms attempts=191633 last=2015-03-31 17:37:26.030673132 -0700 PDT count=11598
	suppress_test.go:43: Error "suppress_test.go:61 #2" at 2015-03-31 17:37:26.330660463 -0700 PDT; delta=99.978889ms attempts=236652 last=2015-03-31 17:37:26.230681213 -0700 PDT count=11087
	suppress_test.go:81: Ran 236652 iterations in 1.100002739s (1.300477321s with wait), fired correctly 51 times (wanted 66) and 21 incorrectly
	suppress_test.go:83: Expected 66 firings, got 51
	suppress_test.go:86: Silencer failed to suppress 21 times
=== RUN TestSilencer5
--- PASS: TestSilencer5 (0.21s)
	suppress_test.go:81: Ran 2279 iterations in 11.002868ms (211.599068ms with wait), fired correctly 12 times (wanted 12) and 0 incorrectly
=== RUN TestSilencerStalled
--- PASS: TestSilencerStalled (0.11s)
	suppress_test.go:104: suppress_test.go:105 anon
	suppress_test.go:104: suppress_test.go:105 anon
FAIL
exit status 1
FAIL	github.com/fastly/go-utils/suppress	5.543s
```

**Comment by jeddenlea** [comment]
Good catch.  I've left the commented tests commented.  They are inherently racy.  But, I've cleaned them up so that they work on my VM reliably with and without a positive GOMAXPROCS.

**Comment by twmb** [comment]
This also removes the minor optimization of not actually starting a goroutine until multiple calls to suppress within the suppress duration have happened.

**Comment by twmb** [comment]
@athomason in terms of the racy tests failing:

In the past, `f` would be scheduled to fire *at least* `suppressTime` apart because suppressing scheduled sleeps, meaning the
```go
if last.IsZero() || now.Sub(last) >= suppressTime {
        firings++
}
```
would always be the case. That is, `now.Sub(last)` would always be at least `suppressTime`.

Now, with the goroutine staying alive until count is zero, and with `f` being called via a timer, `f` does not have to be called at least `suppressTime` apart. The timer could take 101ms on one loop and be fired next at 99ms, autocorrecting itself. The `now.Sub(last) >= suppressTime` check will fail in this racy case.

It seems as if we would want `now.Sub(last) >= suppressTime, +- a few ms of leeway`.

**Comment by jeddenlea** [comment]
Better to perform better under high load than only slightly better under no
load :-)

**Comment by twmb** [comment]
safest code I've ever seen, lgtm.

**Pull request merged**